### PR TITLE
Bluetooth: Controller: Fix Tx Buffer allocation for Encryption Request

### PR DESCRIPTION
**Pull request by cvinayak**
Encryption request is enqueued in thread context from the Tx
buffer pool, so that it is serialized alongwith the already
enqueued data buffers ensuring they are transmitted out to
peer before encryption is setup. Allocate additional Tx
buffers to accomodate simultaneous encryption setup across
active connections.

Samples peripheral_gatt_write and central_gatt_write to
demonstration use of GATT Write Command.

Add BabbleSim test to verify central_gatt_write and
peripheral_gatt_write.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: samples/bluetooth/central_gatt_write/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+sample:
+  name: Bluetooth Central HR
```
Comment (jori-nordic) [inline samples/bluetooth/central_gatt_write/sample.yaml:2]: should this be updated (names n stuff)? same comment for the peripheral yaml

**Review by jori-nordic — CHANGES_REQUESTED** [review]
only need some changes to the yaml files

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 2]
```diff
@@ -196,13 +208,16 @@ static uint8_t force_md_cnt_calc(struct lll_conn *lll_conn, uint32_t tx_rate);
...
 static MFIFO_DEFINE(conn_ack, sizeof(struct lll_tx),
-		    (CONFIG_BT_BUF_ACL_TX_COUNT + CONN_TX_CTRL_BUFFERS));
+		    (CONFIG_BT_BUF_ACL_TX_COUNT + CONN_ENC_REQ_BUFFERS +
+		     CONN_TX_CTRL_BUFFERS));
 
 static struct {
 	void *free;
-	uint8_t pool[CONN_TX_BUF_SIZE * CONFIG_BT_BUF_ACL_TX_COUNT];
+	uint8_t pool[CONN_TX_BUF_SIZE *
+		     (CONFIG_BT_BUF_ACL_TX_COUNT + CONN_ENC_REQ_BUFFERS)];
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c:220]: Given that `CONFIG_BT_BUF_ACL_TX_COUNT + CONN_ENC_REQ_BUFFERS` is used that many places now, would it make sense to add a `#define` for that value, and use that?

File: tests/bluetooth/bsim_bt/bsim_test_gatt_write/src/main.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/util.h>
+
+#include "bs_types.h"
+#include "bs_tracing.h"
+#include "time_machine.h"
+#include "bstests.h"
+
+#define COUNT      5000   /* Arbitrary GATT Write Cmd iterations used */
+#define WRITE_RATE 780800 /* GATT Write bps recorded in this test */
```
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_gatt_write/src/main.c:20]: I assume this is a value measured at some point, and that it should also be able to be calculated as a function with connection interval and write data length provided?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_gatt_write/src/main.c (hunk 3 above)
Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_gatt_write/src/main.c:36]: Added calculations in comment.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: done

File: samples/bluetooth/central_gatt_write/sample.yaml (hunk 1 above)
Comment (cvinayak) [reply samples/bluetooth/central_gatt_write/sample.yaml]: Done

File: samples/bluetooth/central_gatt_write/src/gatt_write_common.c [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+	data_len_max = bt_gatt_get_mtu(conn) - 3;
+	if (data_len_max > BT_ATT_MAX_ATTRIBUTE_LEN) {
+		data_len_max = BT_ATT_MAX_ATTRIBUTE_LEN;
+	}
+
+#if TEST_FRAGMENTATION_WITH_VARIABLE_LENGTH_DATA
+	/* Use incremental length data for every write command */
+	/* TODO: Include test case in BabbleSim tests */
+	static uint8_t direction;
```
Comment (Thalley) [inline samples/bluetooth/central_gatt_write/src/gatt_write_common.c:193]: I am still unsure about this value. What does `direction` reference to? 

It looks like it is being used to either increment or decrement the `data_len`. It is also being used like a boolean value, but is an `uint8_t`. 

Perhaps it should have been `bool decrement;` instead?

**Review by jori-nordic — DISMISSED** [review]

File: samples/bluetooth/central_gatt_write/src/gatt_write_common.c [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+#if TEST_FRAGMENTATION_WITH_VARIABLE_LENGTH_DATA
+	/* Use incremental length data for every write command */
+	/* TODO: Include test case in BabbleSim tests */
+	static bool decrement;
+
+	if (decrement) {
+		data_len--;
+		if (data_len <= 1) {
+			data_len = 1;
+			decrement = 0;
```
Comment (Thalley) [inline samples/bluetooth/central_gatt_write/src/gatt_write_common.c:199]:
```suggestion
			decrement = false;
```

File: samples/bluetooth/central_gatt_write/src/gatt_write_common.c [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+		}
+	} else {
+		data_len++;
+		if (data_len >= data_len_max) {
+			data_len = data_len_max;
+			decrement = 1;
```
Comment (Thalley) [inline samples/bluetooth/central_gatt_write/src/gatt_write_common.c:205]:
```suggestion
			decrement = true;
```

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor request, otherwise LGTM

**Review by Thalley — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Pull request merged**